api/app/geospatial/wps_dataset.py

```diff
@@ -0,0 +1,210 @@
+        if geotransform[0] != other.ds.GetGeoTransform()[0] or geotransform[3] != other.ds.GetGeoTransform()[3]:
+            raise ValueError("The origins of the two rasters do not match.")
+
+        # Create the output raster
+        driver: gdal.Driver = gdal.GetDriverByName("MEM")
+        out_ds: gdal.Dataset = driver.Create("memory", x_size, y_size, 1, self.datatype)
+
+        # Set the geotransform and projection
+        out_ds.SetGeoTransform(geotransform)
+        out_ds.SetProjection(projection)
+
+        self_band: gdal.Band = self.ds.GetRasterBand(self.band)
+        other_band: gdal.Band = other.ds.GetRasterBand(self.band)
+
+        # Process in chunks
+        for y in range(0, y_size, self.chunk_size):
+            y_chunk_size = min(self.chunk_size, y_size - y)
+
+            for x in range(0, x_size, self.chunk_size):
+                x_chunk_size = min(self.chunk_size, x_size - x)
+
+                # Read chunks from both rasters
+                self_chunk = self_band.ReadAsArray(x, y, x_chunk_size, y_chunk_size)
+                other_chunk = other_band.ReadAsArray(x, y, x_chunk_size, y_chunk_size)
+                wider_type = np.promote_types(self_chunk.dtype, other_chunk.dtype)
+
+                self_chunk = self_chunk.astype(wider_type)
+                other_chunk = other_chunk.astype(wider_type)
+
+                other_chunk[other_chunk >= 1] = 1
+                other_chunk[other_chunk < 1] = 0
+
+                # Multiply the chunks
+                self_chunk *= other_chunk
+
+                # Write the result to the output raster
+                out_ds.GetRasterBand(self.band).WriteArray(self_chunk, x, y)
+                self_chunk = None
+                other_chunk = None
+
+        return WPSDataset(ds_path=None, ds=out_ds)
+
+    def warp_to_match(self, other, output_path: str, resample_method: GDALResamplingMethod = GDALResamplingMethod.NEAREST_NEIGHBOUR):
+        """
+        Warp the dataset to match the extent, pixel size, and projection of the other dataset.
+
+        :param ds_to_match: the reference dataset raster to match the source against
+        :param output_path: output path of the resulting raster
+        :param resample_method: gdal resampling algorithm
+        :return: warped raster dataset
+        """
+        source_geotransform = other.ds.GetGeoTransform()
+        x_res = source_geotransform[1]
+        y_res = -source_geotransform[5]
+        minx = source_geotransform[0]
+        maxy = source_geotransform[3]
+        maxx = minx + source_geotransform[1] * other.ds.RasterXSize
+        miny = maxy + source_geotransform[5] * other.ds.RasterYSize
+        extent = [minx, miny, maxx, maxy]
+
+        # Warp to match input option parameters
+        warped_ds = gdal.Warp(
+            output_path,
+            self.ds,
+            options=gdal.WarpOptions(
+                dstSRS=other.ds.GetProjection(),
+                outputBounds=extent,
```

Comment:
Interesting, this was pulled over from https://github.com/bcgov/wps/blob/6b15aa7108601bafde772c1429416f20bda92d42/api/app/utils/geospatial.py#L43 that we currently use.

Reply:
Yeah, I didn't know the issue with the original implementation. I caught it as I was generating those test files. I created the first test file in EPSG4326 which uses lat/lon coordinates and has x and y resolution measured in degrees. I basically used the `warp_to_match` code to then create the EPSG3857 test file but I kept getting errors about not having enough mem/disk space. The problem was that I was taking the x/y res in degress which was very small (0.003333333333067) and passing that in to gdal.Warp which then tries to create an output raster where the x/y resolution is 0.003333333333067 metres! 
Additionally, I ran into problems with the 4326 extent not making sense in the 3857 spatial reference.

Reply:
I'm going to do a quick test and see if specifying a source spatial reference works around the issue.

Reply:
Boo...specifying the `srcSRS` `WarpOption` does help.
ERROR 3: 3005_lats.tif: Free disk space available is 478468026368 bytes, whereas 635088851384008 are at least necessary.

Reply:
Riiiiiight, ya that's an interesting one. Gonna have to think about a solution to that one. Maybe for now we can just implement some sort of check to see if both datasets are in geographic or projected. My brain can't do that right now but maybe we have something like this to use as a check (I don't know if this works, came from chatgippity:

```python
def is_geographic(crs_wkt: str) -> bool:
    """
    Check if the given CRS (WKT string) is geographic (i.e., using degrees).
    :param crs_wkt: The WKT string of the CRS.
    :return: True if geographic, False if projected.
    """
    srs = osr.SpatialReference()
    srs.ImportFromWkt(crs_wkt)
    return srs.IsGeographic()

def get_crs_units(crs_wkt: str) -> str:
    """
    Get the units of the CRS (either 'degrees' for geographic, or 'meters/feet' for projected).
    :param crs_wkt: The WKT string of the CRS.
    :return: The units of the CRS ('degrees', 'meters', 'feet', etc.)
    """
    srs = osr.SpatialReference()
    srs.ImportFromWkt(crs_wkt)
    
    if srs.IsGeographic():
        return 'degrees'
    else:
        return srs.GetLinearUnitsName()
```

Otherwise maybe we have to do some units conversion of meters to degrees? Need to think on that one

Reply:
Oh, I think I figured out one path forward, but we might want to dev chat next week. Right now when this warps, we don't specify the dimensions of the array. We can use `height` and `width` parameters in `WarpOptions` and set those to `self.band.YSize` and self.band.XSize`. Also, we can keep the extent/outputBounds and pass in  `outputBoundsSRS`. Untested as of yet, but my brain is done.

Reply:
My brain is working again (sort of). Specifying the number of pixels for height and width might not be a good idea afterall. I think it could lead to non-square pixels when re-projecting.